test/performance/odp_ipsecfwd.c

```diff
@@ -0,0 +1,1277 @@
+}
+
+static inline odp_bool_t do_ipsec_in(odp_packet_t pkt, const odp_ipsec_sa_t *sa)
+{
+	odp_ipsec_in_param_t param;
+
+	memset(&param, 0, sizeof(param));
+	param.num_sa = 1;
+	param.sa = sa;
+	odp_packet_user_ptr_set(pkt, NULL);
+
+	return odp_ipsec_in_enq(&pkt, 1, &param) == 1;
+}
+
+static inline pkt_handle_result_t handle_ipsec(odp_packet_t pkt, stats_t *stats)
+{
+	odp_ipsec_sa_t *sa = get_in_sa(pkt);
+
+	if (sa == NULL)
+		return PKT_FWD;
+
+	if (odp_unlikely(!do_ipsec_in(pkt, sa))) {
+		++stats->ipsec_in_errs;
+		return PKT_NOK;
+	}
+
+	return PKT_OK;
+}
+
+static inline fwd_entry_t *lookup(odp_packet_t pkt, odph_table_t fwd_tbl)
+{
+	odph_ethhdr_t *eth = odp_packet_l2_ptr(pkt, NULL);
+	uint32_t ip;
+	fwd_entry_t *fwd;
+
+	ip = (odp_be_to_cpu_32)(((odph_ipv4hdr_t *)odp_packet_l3_ptr(pkt, NULL))->dst_addr);
+
+	if (odph_iplookup_table_get_value(fwd_tbl, &ip, &fwd, 0U) < 0 || fwd == NULL)
+		return NULL;
+
+	eth->dst = fwd->dst_mac;
+	eth->src = fwd->pktio->src_mac;
+
+	return fwd;
+}
+
+static inline odp_bool_t forward_packet(odp_packet_t pkt, odph_table_t fwd_tbl, stats_t *stats,
+					int worker_id)
+{
+	fwd_entry_t *fwd = lookup(pkt, fwd_tbl);
+	const pktio_t *pktio;
+
+	if (fwd == NULL) {
+		++stats->discards;
+		return false;
+	}
+
+	pktio = fwd->pktio;
+
+	if (odp_unlikely(pktio->send_fn(pktio, worker_id % pktio->num_tx_qs, pkt) < 0)) {
+		++stats->discards;
+		return false;
+	}
+
+	return true;
+}
+
+static inline void handle_new_packet(odp_packet_t pkt, odph_table_t fwd_tbl, stats_t *stats,
+				     int worker_id)
+{
+	pkt_handle_result_t res = PKT_OK;
+
+	if (odp_unlikely(odp_packet_has_error(pkt))) {
+		++stats->discards;
+		goto err;
+	}
+
+	if (odp_packet_has_udp(pkt))
+		res = handle_udp(pkt, stats);
+	else if (odp_packet_has_ipsec(pkt))
+		res = handle_ipsec(pkt, stats);
+
+	if (odp_unlikely(res == PKT_NOK))
+		goto err;
+
+	if (res == PKT_FWD) {
+		if (!forward_packet(pkt, fwd_tbl, stats, worker_id))
+			goto err;
+
+		++stats->fwd_pkts;
+	}
+
+	return;
+err:
+	odp_packet_free(pkt);
+}
+
+static inline odp_bool_t is_ipsec_in(odp_packet_t pkt)
+{
+	return odp_packet_user_ptr(pkt) == NULL;
+}
+
+static inline void handle_ipsec_packet(odp_packet_t pkt, odph_table_t fwd_tbl, stats_t *stats,
+				       int worker_id)
+{
+	odp_ipsec_packet_result_t result;
+	pkt_handle_result_t res = PKT_FWD;
+
+	if (odp_unlikely(odp_ipsec_result(&result, pkt))) {
+		is_ipsec_in(pkt) ? ++stats->ipsec_in_errs : ++stats->ipsec_out_errs;
+		goto err;
+	}
+
+	if (odp_unlikely(result.status.all != ODP_IPSEC_OK)) {
+		is_ipsec_in(pkt) ? ++stats->ipsec_in_errs : ++stats->ipsec_out_errs;
+		goto err;
+	}
+
+	if (is_ipsec_in(pkt) && odp_packet_has_udp(pkt))
+		res = handle_udp(pkt, stats);
+
+	if (odp_unlikely(res == PKT_NOK))
+		goto err;
+
+	if (res == PKT_FWD) {
+		if (!forward_packet(pkt, fwd_tbl, stats, worker_id))
+			goto err;
+
+		is_ipsec_in(pkt) ? ++stats->ipsec_in_pkts : ++stats->ipsec_out_pkts;
```

Comment:
handle_ipsec_packet() could call is_ipsec_in() only once in the beginning since we do not know how inlineable or fast odp_packet_user_ptr() is in a random ODP implementation. And if the call is done only once, then ip_ipsec_in could be a variable instead of a function.

Reply:
Good point, called only once in v3.